### PR TITLE
[BugOnFeat] Resolve bug with Finetuning

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

The BaseFinetuning Callback contained bugs within its freeze and unfreeze_and_add_param_group functions.

This PR resolves those bugs and add missing tests for those functions.

Furthermore, it renames the callbacks following Callback convention. 

Fixes # (issue) <- this [links related issue to this PR](https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/finetuning.py [hunk 1]
```diff
@@ -49,9 +80,13 @@ def _make_trainable(module: Module) -> None:
         Args:
             module: The module to unfreeze
         """
-        for param in module.parameters():
-            param.requires_grad = True
-        module.train()
+        if isinstance(module, list):
+            for m in module:
+                BaseFinetuning._make_trainable(m)
+        else:
```
Comment (Borda) [inline pytorch_lightning/callbacks/finetuning.py:86]: this is the bug?

File: pytorch_lightning/callbacks/finetuning.py [hunk 2]
```diff
@@ -34,11 +34,42 @@ def multiplicative(epoch):
...
 
 
-class BaseFinetuningCallback(Callback):
+class BaseFinetuning(Callback):
 
     r"""
-    BaseFinetuningCallback.
-    Overrides any functions with your own logic.
+    BaseFinetuning.
+    Overrides ``freeze_before_training`` and ``finetune_function`` with your own logic.
```
Comment (carmocca) [inline pytorch_lightning/callbacks/finetuning.py:41]:
```suggestion
    Override ``freeze_before_training`` and ``finetune_function`` with your own logic.
```

File: pytorch_lightning/callbacks/finetuning.py [hunk 3]
```diff
@@ -106,11 +144,24 @@ def freeze(module: Module, train_bn: bool = True) -> None:
...
                 mod.requires_grad = False
 
+    @staticmethod
+    def filter_on_optimizer(optimizer, params):
+        out_params = []
+        for param in params:
+            is_in = False
+            for group in optimizer.param_groups:
+                for p in group["params"]:
+                    if p.shape == param.shape and torch.equal(p, param):
```
Comment (carmocca) [inline pytorch_lightning/callbacks/finetuning.py:159]:
```suggestion
                    if torch.equal(p, param):
```

File: pytorch_lightning/callbacks/finetuning.py [hunk 4]
```diff
@@ -106,11 +144,24 @@ def freeze(module: Module, train_bn: bool = True) -> None:
...
+                        is_in = True
+            if not is_in:
+                out_params.append(param)
```
Comment (carmocca) [inline pytorch_lightning/callbacks/finetuning.py:162]:
```suggestion
            for group in optimizer.param_groups:
                for p in group["params"]:
                    if torch.equal(p, param):
                        break
            else:
                out_params.append(param)
```

changes:
- Use for..else
- No need to check shape and equals

Comment (carmocca) [inline pytorch_lightning/callbacks/finetuning.py:162]:
```suggestion
            if not any(
                torch.equal(p, param)
                for group in optimizer.param_groups
                for p in group["params"]
            ):
                out_params.append(param)
```

changes:
- Use for..else
- No need to check shape and equals

File: pytorch_lightning/callbacks/finetuning.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:86]: No, the bug was 
```
params = BaseFinetuning.filter_on_optimizer(optimizer, params)
```

**Review by carmocca — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/callbacks/finetuning.py [hunk 5]
```diff
@@ -27,31 +27,67 @@
...
+
+    Example::
+
+        class MyModel(LightningModule)
+
+            ...
+
+            def configure_optimizer(self):
+                # Make sure to filter the parameters based on `requires_grad`
+                return Adam(filter(lambda p: p.requires_grad, parameters))
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/finetuning.py:54]:
```suggestion
                return Adam(filter(lambda p: p.requires_grad, self.parameters))
```

File: pytorch_lightning/callbacks/finetuning.py [hunk 6]
```diff
@@ -27,31 +27,67 @@
...
+
+            def __init__(self, unfreeze_at_epoch=10)
+                self._unfreeze_at_epoch = unfreeze_at_epoch
+
+            def freeze_before_training(self, pl_module):
+                # freeze any module you want
+                # Here, we are freezing ``feature_extractor``
+                self.freeze(pl_module.feature_extractor)
+
+            def finetune_function(self, pl_module, current_epoch, optimizer, opt_idx):
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/finetuning.py:66]:
```suggestion
            def finetune_function(self, pl_module, current_epoch, optimizer, optimizer_idx):
```
can we change it to `optimizer_idx`, just for consistency with wherever it's used in the codebase

File: pytorch_lightning/callbacks/finetuning.py [hunk 7]
```diff
@@ -27,31 +27,67 @@
...
+                # When ``current_epoch`` is 10, feature_extractor will start to be trained.
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/finetuning.py:67]:
```suggestion
                # When ``current_epoch`` is 10, feature_extractor will start training.
```

File: pytorch_lightning/callbacks/finetuning.py [hunk 8]
```diff
@@ -27,31 +27,67 @@
...
     return 2
 
 
-class BaseFinetuningCallback(Callback):
+class BaseFinetuning(Callback):
 
     r"""
-    BaseFinetuningCallback.
-    Overrides any functions with your own logic.
+    BaseFinetuning.
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/finetuning.py:41]:
```suggestion
```
is this required here?

File: pytorch_lightning/callbacks/finetuning.py [hunk 9]
```diff
@@ -27,31 +27,67 @@
...
+    Override ``freeze_before_training`` and ``finetune_function`` with your own logic.
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/finetuning.py:42]:
```suggestion
    Override ``freeze_before_training`` and ``finetune_function`` methods with your own logic.
```

File: pytorch_lightning/callbacks/finetuning.py [hunk 10]
```diff
@@ -27,31 +27,67 @@
...
+
+    .. note:: Make sure to filter the parameters based on `requires_grad`
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/finetuning.py:44]:
```suggestion
    .. note:: Make sure to filter the parameters based on ``requires_grad``.
```

File: pytorch_lightning/callbacks/finetuning.py [hunk 11]
```diff
@@ -27,31 +27,67 @@
...
         """Unfreezes a given module.
         Args:
             module: The module to unfreeze
         """
-        for param in module.parameters():
-            param.requires_grad = True
-        module.train()
+        if isinstance(module, list):
+            for m in module:
+                BaseFinetuning._make_trainable(m)
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/finetuning.py:86]:
```suggestion
                self._make_trainable(m)
```

Comment (rohitgr7) [reply pytorch_lightning/callbacks/finetuning.py:86]: everywhere?

File: tests/callbacks/test_finetunning_callback.py [hunk 12]
```diff
@@ -62,3 +63,55 @@ def on_train_epoch_end(self, trainer, pl_module, outputs):
...
     )
     trainer.fit(model)
+
+
+def test_finetunning_callback_warning(tmpdir):
+    """Test finetunning callbacks works as expected"""
+
+    seed_everything(42)
+
+    class FinetunningBoringModel(BoringModel):
```
Comment (s-rog) [inline tests/callbacks/test_finetunning_callback.py:73]:
```suggestion
def test_finetuning_callback_warning(tmpdir):
    """Test finetuning callbacks works as expected"""

    seed_everything(42)

    class FinetuningBoringModel(BoringModel):
```

File: tests/callbacks/test_finetunning_callback.py [hunk 13]
```diff
@@ -62,3 +63,55 @@ def on_train_epoch_end(self, trainer, pl_module, outputs):
...
+            if epoch == 0:
+                self.unfreeze_and_add_param_group(
+                    pl_module.backbone,
+                    optimizer,
+                    0.1,
+                    train_bn=self.train_bn,
+                    initial_denom_lr=self.initial_denom_lr
+                )
+
+    model = FinetunningBoringModel()
```
Comment (s-rog) [inline tests/callbacks/test_finetunning_callback.py:106]:
```suggestion
    model = FinetuningBoringModel()
```

**Comment by s-rog** [comment]
btw we also got #5643

File: pytorch_lightning/callbacks/finetuning.py [hunk 14]
```diff
@@ -106,26 +144,50 @@ def freeze(module: Module, train_bn: bool = True) -> None:
...
         for mod in module.parameters():
-            if (isinstance(mod, BaseFinetuningCallback.BN_TYPES) and train_bn):
-                BaseFinetuningCallback._make_trainable(mod)
+            if (isinstance(mod, BaseFinetuning.BN_TYPES) and train_bn):
+                BaseFinetuning._make_trainable(mod)
             else:
                 mod.requires_grad = False
 
+    @staticmethod
+    def filter_on_optimizer(optimizer, params):
```
Comment (SkafteNicki) [inline pytorch_lightning/callbacks/finetuning.py:153]:
```suggestion
    def filter_on_optimizer(optimizer: Optimizer, params: Iterable):
```

File: pytorch_lightning/callbacks/finetuning.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:54]: Done !

File: pytorch_lightning/callbacks/finetuning.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:66]: Done !

File: pytorch_lightning/callbacks/finetuning.py [hunk 15]
```diff
@@ -136,16 +198,22 @@ def on_before_accelerator_backend_setup(self, _, pl_module):
...
+        Override to add your freeze logic
+        """
         raise NotImplementedError
 
 
-class BackboneLambdaFinetuningCallback(BaseFinetuningCallback):
+class BackboneFinetuning(BaseFinetuning):
 
     r"""
     Finetunne a backbone model based on a learning rate user-defined scheduling.
```
Comment (SkafteNicki) [inline pytorch_lightning/callbacks/finetuning.py:219]:
```suggestion
    Finetune a backbone model based on a learning rate user-defined scheduling.
```

File: pytorch_lightning/callbacks/finetuning.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:67]: Done !

File: pytorch_lightning/callbacks/finetuning.py [hunk 16]
```diff
@@ -136,16 +198,22 @@ def on_before_accelerator_backend_setup(self, _, pl_module):
...
 
     def freeze_before_training(self, pl_module: LightningModule):
+        """
+        Override to add your freeze logic
+        """
         raise NotImplementedError
 
 
-class BackboneLambdaFinetuningCallback(BaseFinetuningCallback):
+class BackboneFinetuning(BaseFinetuning):
```
Comment (SkafteNicki) [inline pytorch_lightning/callbacks/finetuning.py:216]: Comments to the arg description of this function:
* `verbose` and `should_align` is listed twice
* they should be ordered similar to argument order in the `__init__` function

File: pytorch_lightning/callbacks/finetuning.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:41]: removed !

File: pytorch_lightning/callbacks/finetuning.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:42]: done !

File: pytorch_lightning/callbacks/finetuning.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:44]: done !

File: pytorch_lightning/callbacks/finetuning.py (hunk 11 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:86]: It is a staticmethod.

File: pytorch_lightning/callbacks/finetuning.py (hunk 14 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:153]: done !

File: pytorch_lightning/callbacks/finetuning.py (hunk 15 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:219]: done !

File: pytorch_lightning/callbacks/finetuning.py (hunk 16 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:216]: Good catch !

File: pytorch_lightning/callbacks/finetuning.py (hunk 11 above)
Comment (rohitgr7) [reply pytorch_lightning/callbacks/finetuning.py:86]: we can access it using self. check this: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/callbacks/gpu_stats_monitor.py

**Review by SkafteNicki — APPROVED** [review]
LGTM :]

File: pytorch_lightning/callbacks/finetuning.py [hunk 17]
```diff
@@ -17,166 +17,262 @@
...
-        if not children:
-            if not (isinstance(module, BaseFinetuningCallback.BN_TYPES) and train_bn):
-                for param in module.parameters():
-                    param.requires_grad = False
-                module.eval()
-            else:
-                # Make the BN layers trainable
-                BaseFinetuningCallback._make_trainable(module)
+    def flatten_modules(modules: Union[Module, List[Module]]) -> List[Module]:
+        if isinstance(modules, list):
```
Comment (Borda) [inline pytorch_lightning/callbacks/finetuning.py:87]:
```suggestion
        if isinstance(modules, (tuple, list)):
```
or rather any iterable?

File: pytorch_lightning/callbacks/finetuning.py [hunk 18]
```diff
@@ -17,166 +17,262 @@
...
+            _modules = []
+            for m in modules:
```
Comment (Borda) [inline pytorch_lightning/callbacks/finetuning.py:89]: only one level deep?

File: pytorch_lightning/callbacks/finetuning.py [hunk 19]
```diff
@@ -17,166 +17,262 @@
...
+            if not any(
+                torch.equal(p, param)
+                for group in optimizer.param_groups
+                for p in group["params"]
+            ):
+                out_params.append(param)
+            else:
+                removed_params.append(param)
+
+        if len(removed_params) != 0:
```
Comment (Borda) [inline pytorch_lightning/callbacks/finetuning.py:169]:
```suggestion
        if removed_params:
```

File: pytorch_lightning/callbacks/finetuning.py [hunk 20]
```diff
@@ -17,166 +17,262 @@
...
+            rank_zero_warn(
+                "The provided params to be freezed already exist within another group of this optimizer. "
+                "Those parameters will be skipped. "
+                "HINT: Did you init your optimizer in `configure_optimizer` as such: "
+                f"{type(optimizer)}(filter(lambda p: p.requires_grad, self.parameters()), ...) ", UserWarning)
```
Comment (Borda) [inline pytorch_lightning/callbacks/finetuning.py:174]:
```suggestion
                "The provided params to be freezed already exist within another group of this optimizer."
                " Those parameters will be skipped.\n"
                "HINT: Did you init your optimizer in `configure_optimizer` as such:\n"
                f"{type(optimizer)}(filter(lambda p: p.requires_grad, self.parameters()), ...) ", UserWarning)
```

**Review by Borda — APPROVED** [review]
nice refactoring but not sure what bug it solves...

File: pytorch_lightning/callbacks/finetuning.py [hunk 21]
```diff
@@ -17,166 +17,265 @@
...
+
+            optimizer: The provided optimizer will receive new parameters and will add them to
+                `add_param_group`
+
+            lr: Learning rate for the new param group.
+
+            initial_denom_lr: If no lr is provided, the learning from the first param group will be used
+                and divided by initial_denom_lr.
+
+            train_bn: Wether to train the BatchNormalization layers.
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/finetuning.py:202]:
```suggestion
            train_bn: Whether to train the BatchNormalization layers.
```

File: pytorch_lightning/callbacks/finetuning.py [hunk 22]
```diff
@@ -17,166 +17,265 @@
...
         backbone_initial_lr: Optional, Inital learning rate for the backbone.
             By default, we will use current_learning /  backbone_initial_ratio_lr
+
         should_align: Wheter to align with current learning rate when backbone learning
             reaches it.
+
         initial_denom_lr: When unfreezing the backbone, the intial learning rate will
             current_learning_rate /  initial_denom_lr.
+
         train_bn: Wheter to make Batch Normalization trainable.
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/finetuning.py:265]:
```suggestion
        train_bn: Whether to make Batch Normalization trainable.
```

File: pytorch_lightning/callbacks/finetuning.py (hunk 17 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:87]: Good idea !

File: pytorch_lightning/callbacks/finetuning.py (hunk 18 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:89]: Good catch !

File: pytorch_lightning/callbacks/finetuning.py (hunk 19 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:169]: Done !

File: pytorch_lightning/callbacks/finetuning.py [hunk 23]
```diff
@@ -13,170 +13,292 @@
...
         denom_lr = initial_denom_lr if lr is None else 1.
-        optimizer.add_param_group(
-            {
-                'params': BaseFinetuningCallback.filter_params(module=module, train_bn=train_bn),
-                'lr': params_lr / denom_lr,
-            }
-        )
+        params = BaseFinetuning.filter_params(modules, train_bn=train_bn, requires_grad=True)
+        params = BaseFinetuning.filter_on_optimizer(optimizer, params)
+        if len(params) > 0:
```
Comment (carmocca) [inline pytorch_lightning/callbacks/finetuning.py:233]:
```suggestion
        if len(params):
```

File: tests/base/boring_model.py [hunk 24]
```diff
@@ -93,7 +93,7 @@ def step(self, x):
         return out
 
     def training_step(self, batch, batch_idx):
-        output = self.layer(batch)
+        output = self(batch)
```
Comment (carmocca) [inline tests/base/boring_model.py:96]: Is this necessary for the fix?

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/callbacks/finetuning.py [hunk 25]
```diff
@@ -13,170 +13,292 @@
 # limitations under the License.
 
 r"""
-Finetunning Callback
+finetuning Callback
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/finetuning.py:16]:
```suggestion
Finetuning Callback
```

File: pytorch_lightning/callbacks/finetuning.py [hunk 26]
```diff
@@ -13,170 +13,292 @@
...
+    Override ``freeze_before_training`` and ``finetune_function`` methods with your own logic.
+
+    ``freeze_before_training``: This method is called before ``configure_optimizers``
+        and should be used to freeze any modules parameters.
+
+    ``finetune_function``: This method is called on every train epoch start and should be used to
+        ``unfreeze`` any parameters. Those parameters needs to be added in a new ``param_group``
+        within the optimizer.
+
+    .. note:: Make sure to filter the parameters based on `requires_grad`.
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/finetuning.py:54]:
```suggestion
    .. note:: Make sure to filter the parameters based on ``requires_grad``.
```

File: pytorch_lightning/callbacks/finetuning.py [hunk 27]
```diff
@@ -13,170 +13,292 @@
...
+                if param.requires_grad == requires_grad:
                     yield param
 
     @staticmethod
-    def freeze(module: Module, train_bn: bool = True) -> None:
-        """Freezes the layers up to index n (if n is not None).
+    def make_trainable(modules: Union[Module, Iterable[Union[Module, Iterable]]]) -> None:
+        """
+        Unfreezes the parameters of the provided modules
+        Args:
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/finetuning.py:138]:
```suggestion
        Unfreezes the parameters of the provided modules

        Args:
```

File: .gitignore [hunk 28]
```diff
@@ -147,3 +147,4 @@ test-reports/
 wandb
 .forked/
 *.prof
+cifar-10-python.tar.gz
```
Comment (rohitgr7) [inline .gitignore:150]:
```suggestion
*.tar.gz
```
should we ignore all tar.gz files?

**Review by rohitgr7 — APPROVED** [review]
Looks good. Just few more comments.

File: .gitignore (hunk 28 above)
Comment (tchaton) [reply .gitignore:150]: Yes, we should.

**Pull request merged**